RecoEgamma/EgammaTools/plugins/EGRegressionModifierV3.cc

```diff
@@ -0,0 +1,398 @@
+  //to any electrons with this energy, regardless of their actual eta
+  //hence this lovely hack
+  if(ele.isEB() && maxRawEnergyForLowPtEBSigma_>=0 && 
+     eleRegs_->ecalOnlySigma.useLowEtBin(rawEt,isSaturated)){
+    regData[0] = std::min(regData[0],maxRawEnergyForLowPtEBSigma_);
+  }
+  if(!ele.isEB() && maxRawEnergyForLowPtEESigma_>=0 && 
+     eleRegs_->ecalOnlySigma.useLowEtBin(rawEt,isSaturated)){
+    regData[0] = std::min(regData[0],maxRawEnergyForLowPtEESigma_);
+  }
+  const float ecalSigma = eleRegs_->ecalOnlySigma(rawEt,ele.isEB(),isSaturated,regData.data());
+
+  const float corrEnergy = (rawEnergy + rawESEnergy)*ecalMeanCorr;
+  const float corrEnergyErr = corrEnergy*ecalSigma;
+
+  ele.setCorrectedEcalEnergy(corrEnergy);
+  ele.setCorrectedEcalEnergyError(corrEnergyErr);
+
+  std::pair<float,float> combEnergyAndErr =  eleRegs_->epComb.combine(ele);
+  const math::XYZTLorentzVector newP4 = ele.p4()*combEnergyAndErr.first/ele.p4().t();
+  ele.correctMomentum(newP4, ele.trackMomentumError(), combEnergyAndErr.second);
+}
+
+void EGRegressionModifierV3::modifyObject(pat::Electron& ele) const
+{
+  modifyObject(static_cast<reco::GsfElectron&>(ele));
+}
+
+void EGRegressionModifierV3::modifyObject(reco::Photon& pho) const
+{
+  //check if we have specified an photon regression correction and
+  //return the object unmodified if so
+  if(!phoRegs_) return;
+  
+  const reco::SuperClusterRef& superClus = pho.superCluster();
+
+  // skip HGCAL for now
+  if(superClus->seed()->seed().det() == DetId::Forward ) return;
+
+  // do not apply corrections in case of missing info (happens for some slimmed MiniAOD photons)
+  if(!superClus->clusters().isAvailable()) return; 
+
+  auto regData = getRegData(pho); 
+
+  const float rawEnergy = superClus->rawEnergy(); 
+  const float rawESEnergy = superClus->preshowerEnergy();
+  //bug here, it should include the ES, kept for backwards compat
+  const float rawEt = rawEnergy*superClus->position().rho()/superClus->position().r();
+  const bool isSaturated = pho.nSaturatedXtals();
+  const float ecalMean = phoRegs_->ecalOnlyMean(rawEt,pho.isEB(),isSaturated,regData.data());
+  const float ecalMeanCorr = ecalMean > 0. ? ecalMean : 1.0;
+
+  //see the electrons for explaination of this lovely feature
+  if(pho.isEB() && maxRawEnergyForLowPtEBSigma_>=0 && 
+     phoRegs_->ecalOnlySigma.useLowEtBin(rawEt,isSaturated)){
+    regData[0] = std::min(regData[0],maxRawEnergyForLowPtEBSigma_);
+  }
+  if(!pho.isEB() && maxRawEnergyForLowPtEESigma_>=0 && 
+     phoRegs_->ecalOnlySigma.useLowEtBin(rawEt,isSaturated)){
+    regData[0] = std::min(regData[0],maxRawEnergyForLowPtEESigma_);
+  }
+  const float ecalSigma = phoRegs_->ecalOnlySigma(rawEt,pho.isEB(),isSaturated,regData.data());
+    
+  const double corrEnergy = (rawEnergy + rawESEnergy)*ecalMeanCorr;
+  const double corrEnergyErr = corrEnergy*ecalSigma;
+
+  pho.setCorrectedEnergy(reco::Photon::P4type::regression2, corrEnergy, corrEnergyErr, true);     
```

Comment:
@slava77 , it gets set here.

Reply:
but isn't it followed by https://github.com/cms-sw/cmssw/blob/6e96783e498e1d3ffad261b0bd8d67af2cab1483/RecoEgamma/EgammaPhotonProducers/src/GEDPhotonProducer.cc#L761-L773
?

Reply:
no. That is when you make the gedPhotonTmp collection. The gedPhotons get https://github.com/cms-sw/cmssw/blob/6e96783e498e1d3ffad261b0bd8d67af2cab1483/RecoEgamma/EgammaPhotonProducers/src/GEDPhotonProducer.cc#L918-L921 

This is an absolute mess this code, its almost impossible to figure out what on earth is going on.

Reply:
I mean, this code here is absolutely text book example why I think your code reviews (while sometimes frustrating) are an absolute necessity and huge positive force for good in CMS and probably one of the most important things we can be doing to ensure the long term operation of our experiment